src/databricks/labs/ucx/source_code/notebook_linter.py

```diff
@@ -11,15 +11,31 @@ class NotebookLinter:
     to the code cells according to the language of the cell.
     """
 
-    def __init__(self, languages: Languages):
-        self._languages = languages
+    def __init__(self, langs: Languages, notebook: Notebook):
+        self._languages: Languages = langs
+        self._notebook: Notebook = notebook
+        self._cell_offsets: list[int] = []
 
-    def lint(self, notebook: Notebook) -> Iterable[Advice]:
-        for cell in notebook.cells:
+    @classmethod
+    def from_source(cls, langs: Languages, source: str, default_language: Language) -> 'NotebookLinter':
+        notebook = Notebook.parse("", source, default_language)
+        assert notebook is not None
+        return cls(langs, notebook)
+
+    def lint(self) -> Iterable[Advice]:
+        for cell in self._notebook.cells:
             if not self._languages.is_supported(cell.language.language):
                 continue
             linter = self._languages.linter(cell.language.language)
-            yield from linter.lint(cell.original_code)
+            for advice in linter.lint(cell.original_code):
+                yield advice
+                self._cell_offsets.append(cell.original_offset)
```

Comment:
Shouldn't it be the first statement in the loop? Otherwise there's a bug, where offsets are counted only for the bad cells

Reply:
I only count them for bad cells. It is the only way to correlate the offsets wihtout storing the cell number in the advice. HOwever, we have now changed the requirement to just report the notebook source line anyway, so this will now become irrelevant

Reply:
No longer relevant